### PR TITLE
Extracting common abstract base class for transition info

### PR DESCRIPTION
**Pull request by HenningNT**
Extracting common abstract base class for transition info.
Added new abstract base class for common properties.
Updating tests and other code as needed.

### CODE REVIEW

**Comment by HenningNT** [comment]
... as mentioned in #143

File: src/Stateless/TransitionBase.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
+﻿namespace Stateless
```
Comment (nblumhardt) [inline src/Stateless/TransitionBase.cs]: Looks like this file might be a leftover?

File: src/Stateless/Reflection/TransitionInfo.cs [hunk 2]
```diff
@@ -1,52 +1,18 @@
...
-        /// </summary>
-        public StateInfo DestinationState { get; private set; }
-
+        public TriggerInfo Trigger { get; internal set; }
 
         /// <summary>
         /// Description of provided guard clause, if any.
         /// </summary>
-        public string GuardDescription { get; private set; }
+        public string GuardDescription { get; internal set; }
```
Comment (nblumhardt) [inline src/Stateless/Reflection/TransitionInfo.cs]: Could this be a `protected set`?

File: src/Stateless/TransitionBase.cs (hunk 1 above)
Comment (HenningNT) [inline src/Stateless/TransitionBase.cs]: Yes, that should be there... It doesn't show up in Solution Explorer, I'll remove this!

**Comment by nblumhardt** [comment]
Looks good!

I'm a little uneasy with the name `ExternalTransitions` - but I realise you're sticking with the standard terminology and I can't come up with a better alternative. It's really odd how (at least in [the UML lexicon](https://en.wikipedia.org/wiki/UML_state_machine#Local_versus_external_transitions)), _external_ transitions are the opposite of _local transitions_, rather than _internal transitions_ (which also appear in the lexicon). Eek! :-)

Any other possible options for the name? If not, this all seems ready to go.

File: src/Stateless/Reflection/TransitionInfo.cs (hunk 2 above)
Comment (HenningNT) [inline src/Stateless/Reflection/TransitionInfo.cs]: Yes, it can :-)

**Comment by HenningNT** [comment]
Naming things are hard...
I just had a look at the latest UML specification:

> 14.2.3.8.1 Transition kinds relative to source
> The semantics of a Transition depend on its relationship to its source Vertex. Three different possibilities are defined,
> depending on the value of the Transition’s kind attribute:
> • kind = external means that the Transition exits its source Vertex. If the Vertex is a State, then executing this
> Transition will result in the execution of any associated exit Behavior of that State.
> • kind = local is the opposite of external, meaning that the Transition does not exit its containing State (and,
> hence, the exit Behavior of the containing State will not be executed). However, for local Transitions the target
> Vertex must be different from its source Vertex. A local Transition can only exist within a composite State.
> • kind = internal is a special case of a local Transition that is a self-transition (i.e., with the same source and
> target States), such that the State is never exited (and, thus, not re-entered), which means that no exit or entry
> Behaviors are executed when this Transition is executed. This kind of Transition can only be defined if the
> source Vertex is a State.

I think the names are fairly OK, and also realize that we should maybe rename InternalTransition() to something else, maybe PermitInternal() and add PermitLocal() ;-)

**Comment by nblumhardt** [comment]
Thanks for the follow-up. There is some appeal in renaming `InternalTransition` to `PermitInternal()`, or something along those lines.

I think `ExternalTransition` is actually a misnomer now, since because of the implicit hierarchy, this will also include local transitions. (I guess this is what you're getting at with `PermitLocal()`).

I have to run, but I'll give it some more thought!

**Comment by HenningNT** [comment]
Something along those lines, yes. If we want to conform to the UML naming, then we could either use PermitExternal, PermitLocal and PermitInternal. Or just go with Permit, and let the Permit method figure it out. The latter has a simpler API, but having explicit Permits makes it quite clear to the users what they are asking for.

For the reflection stuff it might be worth splitting up transitions into ExternalTransition, LocalTransition and InternalTransition?

**Comment by nblumhardt** [comment]
I don't think the switch to `PermitExternal`/`PermitLocal` would work so well with the current API, since the difference between the two is implied by the state hierarchy. Maybe there's too much water under the bridge to be fully aligned with the UML terms.

From the model's point of view, would it make sense to say we have two kinds of transitions - those we know the destination state of (...fixed...?) and those we don't (dynamic, today). I think the reflection end of things we could have `TransitionInfo` (abstract), `FixedTransitionInfo` and `DynamicTransitionInfo`?

Then at the API, we just try to fit `InternalTransition` in among the other `Permit` variants:

```csharp
    .Permit(Trigger.Closed, State.Closed)              // FixedTransitionInfo
    .PermitDynamic(Trigger.Closed, t => State.Closed)  // DynamicTransitionInfo
    .PermitReentry(Trigger.Assigned)                   // FixedTransitionInfo
    .PermitInternal(Trigger.Assigned, t => Act())      // FixedTransitionInfo
```

(I think this is a breaking change we could afford to take in the new major version.)

The alternative would be to model internal transitions as "actions", so `PermitAction()` or just `OnTrigger()`. I'm okay with the first, would need to think more carefully about the second.

What do you think?

**Comment by HenningNT** [comment]
I like the Dynamic/Fixed TransitionInfo! I'll try that, and see what happens :-)

We can discuss changing the API later 👍

**Comment by HenningNT** [comment]
ok, so I tried to simplify it a bit... maybe I did too many changes?

File: test/Stateless.Tests/DotGraphFixture.cs [hunk 3]
```diff
@@ -33,7 +33,7 @@ public void SimpleTransition_DotGraph()
 
             sm.Configure(State.A)
                 .Permit(Trigger.X, State.B);
-
+            var resukt = DotGraphFormatter.Format(sm.GetInfo());
```
Comment (nblumhardt) [inline test/Stateless.Tests/DotGraphFixture.cs]: This doesn't look like it belongs here :-)

**Comment by nblumhardt** [comment]
Great, thanks for the update! Still trying to get some time spare to run back through it :-)

File: test/Stateless.Tests/DotGraphFixture.cs (hunk 3 above)
Comment (HenningNT) [inline test/Stateless.Tests/DotGraphFixture.cs]: You're right!

File: src/Stateless/Reflection/StateInfo.cs [hunk 4]
```diff
@@ -130,12 +122,7 @@ public class StateInfo
         /// <summary>
         /// Transitions defined for this state.
         /// </summary>
-        public IEnumerable<TransitionInfo> Transitions { get; private set; }
-
-        /// <summary>
-        /// Transitions defined for this state internally.
-        /// </summary>
-        public IEnumerable<TransitionInfo> InternalTransitions { get; private set; }
+        public IEnumerable<FixedTransitionInfo> ExternalTransitions { get; private set; }
```
Comment (nblumhardt) [inline src/Stateless/Reflection/StateInfo.cs]: Should/could this be `FixedTransitions`?

File: src/Stateless/Reflection/StateInfo.cs [hunk 5]
```diff
@@ -130,12 +122,7 @@ public class StateInfo
         /// <summary>
         /// Transitions defined for this state.
         /// </summary>
-        public IEnumerable<TransitionInfo> Transitions { get; private set; }
```
Comment (nblumhardt) [inline src/Stateless/Reflection/StateInfo.cs]: Could this remain, but simply return `FixedTransitions.Concat(DynamicTransitions)`? In .NET reflection, we have this pattern in a few places e.g. https://msdn.microsoft.com/en-us/library/system.reflection.typeinfo.declaredmembers(v=vs.110).aspx)

File: src/Stateless/Reflection/StateInfo.cs (hunk 4 above)
Comment (HenningNT) [inline src/Stateless/Reflection/StateInfo.cs]: Yup!

File: src/Stateless/Reflection/StateInfo.cs (hunk 5 above)
Comment (HenningNT) [inline src/Stateless/Reflection/StateInfo.cs]: Sure we can, some might find it useful.

**Comment by nblumhardt** [comment]
Cool :-) ... good to go? 🚢 🚢 🚢

**Comment by HenningNT** [comment]
I think so! 👍

**Comment by nblumhardt** [comment]
Great! :-)

Circling back... What do you think about a possible rename from `InternalTransition()` to `PermitSomething()`?

**Comment by HenningNT** [comment]
Renaming `InternalTransition `to `PermitInternal `seems like a sensible move.

**Pull request merged**